worker/src/handles/TcpConnection.cpp

```diff
@@ -108,7 +108,7 @@ void TcpConnection::Close()
 		err       = uv_shutdown(
       req, reinterpret_cast<uv_stream_t*>(this->uvHandle), static_cast<uv_shutdown_cb>(onShutdown));
 
-		if (err != 0)
+		if (err != 0 && err != UV_ENOTCONN)
```

Comment:
Why this?

Reply:
Sorry, I should have probably skipped this here.

Basically I couldn't figure out why it ends up with this error instead of shutting down properly in Rust version no matter the order things are happening in. It never sets `isClosedByPeer` to `true` since `Close()` is called directly. But I though that if on shutdown connection is already dropped, this is not something that is worth killing the process.

It wasn't an issue when worker was a process, but it is when running in a thread since it takes the whole app with it.

I might be understanding it wrong, please correct me if that is the case.

Reply:
Ok, we just call `MS_ABORT` when something should NOT happen at all (or we think that it should never happen). Isn't this true? is it legitimate for this to happen? when?

Reply:
May be `uv_shutdown()` can exit with `UV_ENOTCONN` if the process is closed before it completes?

Reply:
I will need to dig into `uv_shutdown()` sources before I can answer, docs didn't have explanation as to why this may happen.

Reply:
So the code looks like this:
```c
int uv_shutdown(uv_shutdown_t* req, uv_stream_t* stream, uv_shutdown_cb cb) {
  assert(stream->type == UV_TCP ||
         stream->type == UV_TTY ||
         stream->type == UV_NAMED_PIPE);

  if (!(stream->flags & UV_HANDLE_WRITABLE) ||
      stream->flags & UV_HANDLE_SHUT ||
      stream->flags & UV_HANDLE_SHUTTING ||
      uv__is_closing(stream)) {
    return UV_ENOTCONN;
  }
```

And in my case it is `!(stream->flags & UV_HANDLE_WRITABLE)`. Why it is not writable I'm really not sure as of right now.

Reply:
Hm... maybe it should have `this->role == UnixStreamSocket::Role::PRODUCER` in the condition there?

Reply:
Yep, for some reason it tries to call that on reading end of the pipe, which according to the `uv_shutdown()` should fail with `UV_ENOTCONN`. How does it not reach that place with Node.js library I don't know.

Applied what I believe is the correct patch in the latest commit.